src/node_crypto.cc

```diff
@@ -3174,26 +3245,54 @@ void CipherBase::SetAuthTag(const FunctionCallbackInfo<Value>& args) {
 }
 
 
-bool CipherBase::SetAAD(const char* data, unsigned int len) {
+bool CipherBase::SetAAD(const char* data, unsigned int len, int plaintext_len) {
   if (ctx_ == nullptr || !IsAuthenticatedMode())
     return false;
+
   int outlen;
-  if (!EVP_CipherUpdate(ctx_,
-                        nullptr,
-                        &outlen,
-                        reinterpret_cast<const unsigned char*>(data),
-                        len)) {
-    return false;
+  const EVP_CIPHER* const cipher = EVP_CIPHER_CTX_cipher(ctx_);
+  int mode = EVP_CIPHER_mode(cipher);
+
+  // When in CCM mode, we need to set the authentication tag and the plaintext
+  // length in advance.
+  if (mode == EVP_CIPH_CCM_MODE) {
+    if (plaintext_len < 0) {
+      env()->ThrowError("plaintextLength required for CCM mode with AAD");
+      return true;
```

Comment:
I admit, this was a bit hacky. Primarily to avoid unnecessarily setting the return value here:

```c++
if (!cipher->SetAAD(Buffer::Data(args[0]), Buffer::Length(args[0]), pt_len))
  args.GetReturnValue().Set(false);  // Report invalid state failure
```